---
title: "seqdenoise-vignette"
author: "Cameron M. Nugent"
data: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{seqdenoise-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, echo = FALSE}
library(seqdenoise)
```

## Abstract


## Introduction
preamble here

### Implementation

### Dependencies
`seqdenoiser` is dependent on [the `aphid` package](https://github.com/shaunpwilkinson/aphid) for comparison of sequences against the COI-5P PHMMs. The `ape` is also a requirement because sequences are internally converted to ape "DNAbin" objects. 


### The data

For demonstration of the workflow, `seqdenoise` contains several example files. The first is a fastq file with series of single molecule, real-time (SMRT) sequencing reads generated on the Pacific BioSciences SEQUEL platform. These data are derived from 5 unique samples animal samples, from which ~658bp amplicons of the mitochondrial cyctochrom c oxidase I gene were isolated and sequenced. 

`seqdenoise` is able to operate on inputs from either fasta or fastq files, or their gzipped equivalents. 

```{r}

fastq_example_file = system.file('extdata/sequel_smrt_subset.fastq', package = 'seqdenoise')

fastq_example_file = system.file('extdata/sequel_smrt_subset.fastq.gz', package = 'seqdenoise')

```


## Using the package
### for the impatient
The progam contains a wrapper function `denoise_file` that allows for a user to run the entire denoiser pipeline in a single step (there are quantitatively informed defaults for executing the full pipline, but it is highly reccommended that a user acclimate themselves to the `denoise` pipeline options to ensure that the paramaters make sense for their given dataset and needs)

All that is needed is for the input and output files to be specified. The program by default will read in the input file, run each sequence through the denoising algorithm, and then output the denoised sequences to the specified output file. The denoise pipeline contains certain rules (see below for details on these paramaters) that lead to some sequences being rejected entirely, the `keep_rejects` option allows a user to output these rejected sequences to another file for futher inspection. The program also has an option for generating a simple log file, and allows for multicore execution of the denoising pipeline, by simply specify the number of avaliable cores.
```{r, eval = FALSE}
#NOTE - this code is not run so as to avoid the generation of output files!
#
#our input file with noisy sequence data is fastq_example_file. 
#output is "example_output.fastq"
denoise_file(fastq_example_file, filename = "example_output.fastq")

denoise_file(fastq_example_file,filename = "multicore-example_output.fastq", multicore = 8, log_file = TRUE, keep_rejects = TRUE )


 
```


### batch analyses in R

```{r}

data = read_fastq(fastq_gz_real_file)
head(data)

```

```{r}
dim(data)
```

```{r}
x = denoise(data$sequence[[71]], keep_phred = FALSE)
x

```

```{r}
x = denoise(data$sequence[[78]], phred = data$quality[[78]])
x

```

```{r}
x = DNAseq(data$sequence[[71]], name = 'test1')

```

```{r}

test = denoise(data$sequence[[71]], to_file = FALSE)
```

```{r}
outputs = list()
for(i in 1:15){
  print(i)
  #print(data$sequence[i])
  out = denoise(data$sequence[i], name = data$header_data[i], phred = data$quality[i], to_file = FALSE )
  outputs[[i]] = out
}

print("output length:")
length(outputs)
```


```{r}
outputs[[2]]
```





